### PR TITLE
[RDY] Fix: deadlock when switching modes

### PR DESCRIPTION
**Pull request by ramirezmike**
see #228

### CODE REVIEW

**Comment by TimUntersberger** [comment]
That is a very good solution in my opinion!

**Comment by ramirezmike** [comment]
So I added 9fdd7fa to address the other deadlock in #228 
And the last commit is just formatting changes.

For the fix for 9fdd7fa I did a couple similar changes to the first commit: changed locks to try locks on paths that are acceptable to fail. In this case, it will potentially fail on updating the appbar (which, again, just will refresh like 20ms later) OR it will ignore a key event.. but, this would only happen when we're switching keybindings anyway. 

The other piece is I did swap the SendMessageA with a PostMessageA when the appbar window gets closed. This is because the SendMessageA blocks the thread it comes from _and_ the thread that handles the close (the same thread that paints every 10ms or whatever). 

There's more information [here](https://docs.microsoft.com/en-us/windows/win32/winmsg/about-messages-and-message-queues#message-deadlocks) but basically if you have a thread that sends a SendMessageA and another thread that calls PeekMessage before checking if there was a SendMessage, you'll deadlock. The docs suggest calling InSendMessage before PeekMessage and then handling the event prior to calling PeekMessage to avoid a deadlock, but I had trouble getting that to work right.

Using PostMessageA let's a draw event finish before then closing the appbar and prevents that deadlock.


I tested this by just going crazy switching modes, toggling appbar and taskbar and seeing if it deadlocked. It cleared up those issues for me HOWEVER I think there are two keybinding issues that still exist which I'll make a separate issue for.

@TimUntersberger I'd recommend testing this out just to make sure this behaves the same on both our setups. Also, I have this set for development right now but idk if you'd want it in master instead? Not sure if you're still debating on the existence of the development branch.

**Comment by ramirezmike** [comment]
I realized that I should probably explain the keybinding issues briefly so you know what to look out for:

**Keybinding issue 1**

This one I think is minor and I actually haven't reproduced it I just suspect it's an issue from seeing how the code behaves. In the hot reload code it fires an event to unregister all the current keybindings and then updates the keybindings to whatever is in the new config with set_keybindings. However, I consistently saw the set finishes mutating the KB manager's bindings before the unregister keybindings event is caught. This could mean that if a user modified their config's keybindings and did a hot reload, the unregister keybindings event would unregister their new set of keybindings and not their old one. Again, didn't test this but I suspect this happens.

**Keybinding issue 2**

This one I did run into and has to do with changing modes. I don't know what the exact steps are yet but occasionally I'd end up with a mode activated but no keybindings registered. This appears at first like a deadlock because nothing seems to respond but you can tell it's not a deadlock by checking a couple things

1) moving the mouse cursor over the appbar shouldn't show the Windows busy cursor. If it was a deadlock, the cursor would change indicating the window is busy because the operating system is detecting that the movemouse events it's sending to nog aren't being handled

2) if you have a time with seconds on your app bar you should see it continue to tick

3) if you have multiple programs managed in the current workspace _and_ you have the currently focused application component on your appbar you should see it update as you manually change focus with alt+tab or clicking on your different programs.

Basically, if you find nog unresponsive but the appbar looks like it's updating, then it's a keybinding issue not a deadlock issue.

**Comment by TimUntersberger** [comment]
>  Not sure if you're still debating on the existence of the development branch.
I want to delete development asap, but I don't know when a good time is because we would have to update all of the branches right? I never had to do this so Idk how long this would take.

Have you considered using [SendNotifyMessage](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-sendnotifymessagea)?

**Comment by ramirezmike** [comment]
> Have you considered using [SendNotifyMessage](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-sendnotifymessagea)?

That's a good middle ground behavior of PostMessageA and SendMessageA. I suspect it will behave well.

> 
> > Not sure if you're still debating on the existence of the development branch.
> 
> I want to delete development asap, but I don't know when a good time is because we would have to update all of the branches right? I never had to do this so Idk how long this would take.

Well, we could just start making pull requests against master and ignore the development branch and then at some point it would be easier to delete, right?

**Comment by ramirezmike** [comment]
Ok, I rebased on master and added two commits

26ca29e - SendNotifyMessage works so I changed it to that
edfbee2c - Figured out the issue with mode toggling keybinding not working occasionally. Turns out that the set_keybindings function doesn't set the mode_handlers from the new config when hot reloading. This causes a panic if the user triggers a hot reload of the config then attempts to use a mode that they hadn't used before triggering the hot reload. 

That is, if you go through all your modes (and "execute" each mode handler) before triggering the hot_reload, everything works fine. But, if you hot_reload before triggering your modes, you end up with an empty mode_handlers and then nog crashes when you try to switch modes because it doesn't find the mode handler to execute.

So I fixed that by just having the hot_reload also pass in the mode_handlers from the newly parsed config and the panics went away.

I think this resolves all the big issues that I see and think this is a good release candidate if we merge it into master. Feel free to merge it whenever works for you.

**Comment by TimUntersberger** [comment]
I'd see after fixing `Deadlock B` (https://github.com/TimUntersberger/nog/issues/228#issuecomment-781701920) we merge this and create a new release.

**Comment by ramirezmike** [comment]
Yeah, just working on trying to reproduce it now...

**Comment by ramirezmike** [comment]
Alright, looks like this is good to go so merging

**Comment by TimUntersberger** [comment]
@ramirezmike created a new release `v0.13.0` based of master (Merged with development).
